### ISSUE TITLE
unrelated type checks on records should support pairwise assignability

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#4438]
**Describe the issue**

I think "pairwise" is the right term. Anyhow, the lint will report on the following code:

```dart
var list = <(int? a, int? b)>[];
list.contains((null, 7));
```

because the argument type is `(Null, int)`, which is not exactly equal to `(int? int?)`. Bummer.

I think the logic should be:

* if the two types are each record types,
* and they are the same shape,
* and for each field F in the first record type and the corresponding field G in the second record type, F is assignable to G or G is assignable to F (or maybe use linter's notion of "related", so: F and G are related (should be symmetric),
* then they are not unrelated.

**Comment by lrhn** [comment]
I think a better word would be "pointwise". That works even for non-pairs.

And the "is related type" should probably be pointwise "is related type" on the field types of records with the same shape.

**Comment by eernstg** [comment]
I'm surprised that two types `S` and `T` are considered unrelated in the situation where `S <: T`. However, I don't get a lint for the following:

```dart
void main() {
  var list = <(int? a, int? b)>[];
  print(list.contains((null, 7)));
}
```

with an 'analysis_options.yaml' as follows:

```yaml
linter:
  rules:
    - collection_methods_unrelated_type
```

However, I _do_ get the lint if `7` is replaced by `7.5`. So it's not obvious how we'd have this:

> the lint will report on the following code ...

Maybe it has already been fixed?

**Comment by srawlins** [comment]
Oops, yes I think this is fixed already. I think I had an unrelated issue in the nullability of the values in my example. I'll close for now.
